[WebNN EP] Support Einsum op

peishenyan opened the pull request:
Adds support for einsum via WebNN matmul, transpose, reshape, reducesum, identity and element-wise binary ops.

Honry commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,508 @@
+  case RecognizedOperatorType::Identity:
+    {
+      emscripten::val input = model_builder.GetOperand(node.InputDefs()[0]->Name());
+      output = model_builder.GetBuilder().call<emscripten::val>("identity", input);
```
Please note that `identity` has not been supported by XNNPack backend, @lisha0314 will help implement it in Chromium soon.

fdwr replied:
Note you can also just map `identity` to `transpose`, if XNNPack supports transpose (since a transpose with sequential axes is a nop). I don't mind either way - it's fine to just wait for XNNPack.

peishenyan replied:
Maybe I just need to pass the input to the output? 🤣 I think it is unnecessary to call `identity`
```
emscripten::val input = model_builder.GetOperand(node.InputDefs()[0]->Name());
output = input;
```

Honry commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,508 @@
+      auto kept_axes = m_components.back().GetLabels(m_label_indices);
+      assert(kept_axes.size() <= 1);
+      std::vector<uint32_t> reduced_axes;
+      // uint32_t kept_axes_mask = (kept_axes.size() << 0) - 1;
```
Pls. remove this unused line.

Honry commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,508 @@
+                                        const logging::Logger& logger) const {
+
+  if (device_type == WebnnDeviceType::CPU) {
+    LOGS(logger, VERBOSE) << "Einsum is not supported for cpu in WebNN EP. Matmul and ReduceSum are not supported in XNNPACK.";
```
`Matmul` is just implemented in Chromium, we should only restrict `recognized_operator_type == RecognizedOperatorType::ReduceSum`

Honry commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,508 @@
+  }
+
+  /* // dml's condition, im not sure if it is necessary
+  if (static_cast<uint32_t>(node.OutputDefs()) != 1) {
```
Pls. remove these commented codes.

Honry commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,508 @@
+          }
+      }
+
+      emscripten::val input = model_builder.GetOperand(node.InputDefs()[0]->Name());
```
Maybe we could centralize the first input outside the `switch` to simplified the code.

peishenyan replied:
`PairwiseOperandProcess` needs to get two inputs in function. Centralizing the first input outside the `switch` may break its "integrity". I prefer to decentralize them.

Honry reviewed (commented):
Thanks @peishenyan, some comments.

peishenyan commented:
Thanks @Honry , comments addressed.
@fdwr , @guschmue , PTAL, thanks!

peishenyan commented:
Hi, all. Last week, I implemented Einsum based on the DML execution provider's implementation, which only support a few conditions. However, Einsum('bhwc,hkc->bhwk', A, B) cannot be handled in ``Segment Anything Encoder''. 

Now, according to Einsum Formula in [ML Operator Formulas](https://fdwr.github.io/MachineLearningOperators/OperatorFormulas.html), I implement pair-wise operand process for handling two inputs and kept the previously implemented single operand process (Identity, Transpose and ReduceSum.)

peishenyan commented:
Since WebNN does not support Triangular op so far and Diagonal depends on triangular, WebNN will fallback when the Einsum op is used for Diagonal operation.

One more thing, in Segment Anything Encoder model, the output shape of Einsum is not defined, which is the input of some Unsqueeze operation, while WebNN does not support dynamic shape. We need additional work to solve this problem in another work branch.

peishenyan commented:
I will leave for about one week for my research paper submission. Responses may be delayed.

Honry commented:
Thanks @peishenyan! 

@guschmue, @fdwr, PTAL thanks!

> One more thing, in Segment Anything Encoder model, the output shape of Einsum is not defined, which is the input of some Unsqueeze operation, while WebNN does not support dynamic shape. We need additional work to solve this problem in another work branch.

@guschmue, @fdwr, the output shape of `Einsum` is not provided in the ONNX graph even it is a static shape model. You know its output shape could only be calculated via its `equation` expression and input shape. 

So my question is, is that possible to calculate and add its output shape info to the graph during ORT Web graph optimization?

Honry commented:
Friendly ping @jgraham, @foolip, anyone can help on the CI failures? Looks like it is irrelevant to this PR.

Honry commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,601 @@
+                                        const WebnnDeviceType device_type,
+                                        const logging::Logger& logger) const {
+  emscripten::val console = emscripten::val::global("console");
+  console.call<void>("log", emscripten::val("log from Einsum..."));
```
Please remove this debug code.

fdwr commented:
> So my question is, is that possible to calculate and add its output shape info to the graph during ORT Web graph optimization?

Are you saying that ORT is not doing shape inference properly for EinSum? The DML EP (which also needs constant shapes before DML CompileGraph is called) registers its own shape inference function for EinSum here https://github.com/microsoft/onnxruntime/blob/1e69b612382205b0588f08d2b808b12e32a50a51/onnxruntime/core/providers/dml/OperatorAuthorHelper/OperatorHelper.cpp#L1632-L1641, which (if you've already parsed the components) isn't *too* long/complicated. 

> Since WebNN does not support Triangular op so far and Diagonal depends on triangular, WebNN will fallback when the Einsum op is used for Diagonal operation.

👍 Lisha is working on it (and I drew some whiteboard diagrams on decomposing it for DML feature level 4). It should be available soon.

> ... my research paper submission

Thanks for creating this - hope your research paper goes well.

Honry commented:
> Are you saying that ORT is not doing shape inference properly for EinSum? 

Right, ORT doesn't do shape inference for `EinSum`, we couldn't get the input shape of `EinSum` node's following nodes from the graph, e.g. in following snapshot, we couldn't get the `Usqueeze` node's input shape. Thus this op will be fallback to wasm as we do dynamic input shape check at https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/webnn/builders/helper.cc#L73.

<img width="122" alt="image" src="https://github.com/microsoft/onnxruntime/assets/3271201/513a06ae-bdde-46e1-ae8c-a6588d8039f6">

> The DML EP (which also needs constant shapes before DML CompileGraph is called) registers its own shape inference function for EinSum here

Interesting, how and when does DML EP register the shape info to the graph?

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,508 @@
+
+// Helper functions, thanks for DML OperatorHelper
+enum class RecognizedOperatorType
+{
```
Yep, while putting both braces symmetrically at the same indentation level has readability benefits for faster eye scanning, ORT prefers adenserformat, and so consistency matters more.

Honry commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,601 @@
+// Add operator related.
+
+void EinsumOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  model_builder.AddInitializerToSkip(node.InputDefs()[1]->Name());
```
Don't skip the input as we need to register it to a WebNN constant.

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,598 @@
+      emscripten::val a = model_builder.GetOperand(node.InputDefs()[a_idx]->Name());
+      emscripten::val b = model_builder.GetOperand(node.InputDefs()[b_idx]->Name());
+      output = model_builder.GetBuilder().call<emscripten::val>("mul", a, b);
+    } break;
```
🔎👀 Dense code is tougher to read. Inserting some line breaks between the case statements would help readers logically see the separate blocks better.

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,598 @@
+      std::vector<int32_t> axes_data;
+      std::transform(
+          reduced_axes.begin(), reduced_axes.end(), std::back_inserter(axes_data),
+          [input_rank](int64_t axis) -> int32_t { return SafeInt<int32_t>(HandleNegativeAxis(axis, input_rank)); });
```
Hmm, didn't @honry already add a helper function for this really common downcasting pattern?

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,598 @@
+    } break;
+    case RecognizedOperatorType::Transpose: {
+      emscripten::val input = model_builder.GetOperand(node.InputDefs()[0]->Name());
+      // Transpose via input strides. The output tensor is not strided.
```
This comment doesn't make sense in the current context, since strides are not used in the WebNN EP.

fdwr replied:
This comment doesn't make sense in the current context, since strides are not used in the WebNN EP.

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,598 @@
+                              emscripten::val& output,
+                              const logging::Logger& logger) {
+  auto input_a_labels = m_components[0].GetLabels(m_label_indices);
+  auto input_b_labels = m_components[1].GetLabels(m_label_indices);
```
An assert would be wise, as we don't want PairwiseOperandProcess to read `m_components[1]` and `m_components[2]` before ensuring `m_components.size() == 3`.

peishenyan replied:
Considering that the PairwiseOperandProcess function is only triggered when `recognized_operator_type==RecognizedOperatorType::PairWise`, this ensures that there are three elements in m_components. Therefore, it might not be necessary to include an assert statement in this context?

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,598 @@
+  ReduceSum,
+  Transpose,
+  Multiply,
+  Others,
```
`Others` implies it handles all others, but it only handles binary inputs (PairwiseOperandProcess). Propose a clearer name like `BinaryInputs` or `PairwiseInputs`.

fdwr reviewed (commented):
Thanks Peishen. I'm still trying to figure out `PairwiseOperandProcess` (some more comments/visuals/examples could help, especially future generations), but I reviewed the rest of it.

peishenyan commented:
Hi @fdwr , I have addressed most of the comments, but the code is not ready for final review now. 
As @Honry said before, the output shape of Einsum is not provided in the ONNX graph. We found that we could use onnx shape inference to achieve almost all output shapes except the output shape of Einsum. Since there was only Rank Inference function for Einsum instead of Shape Inference, we implemented Shape Infernce for Einusm in https://github.com/onnx/onnx/pull/6010 to address this problem, which is still waiting for review and merge.
Furthermore, I am working on supporting parsing ellipsis in equation and diagnoal now~

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -36,7 +36,7 @@ enum class RecognizedOperatorType {
   Transpose,
   Multiply,
-  Others,
+  PairWise,
```
```suggestion
  Pairwise
```
(compound word)

peishenyan commented:
Hi  @Honry @fdwr , I have addressed all the comments and we decide to implement the extended features (diagonal and ellipsis parsing) in the future, which will not be included in this PR. This PR is ready for review. PTAL. Thanks

tianleiwu commented:
/azp run Windows ARM64 QNN CI Pipeline,Windows x64 QNN CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline

tianleiwu commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models

tianleiwu commented:
Any test can cover the op in WebNN EP?

Honry commented:
> Any test can cover the op in WebNN EP?

@tianleiwu, good point!

@peishenyan has ever tested the following unit tests, but which are all fp64 data type that currently is not supported by WebNN. Is that possible to change the data type to fp32?
https://github.com/microsoft/onnxruntime/blob/main/js/web/test/suite-test-list.jsonc#L1731-L1735

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,613 @@
+
+  std::vector<uint32_t> sequence_o(output_indices.size());
+  std::iota(sequence_o.begin(), sequence_o.end(), 0);
+  if (v != sequence_o) {
```
```
  if (output_permutation != sequence_o) {
```
We should use a clearer name than just "v". With the common exception of `i` for loop counters and `x`,`y`,`z` for coordinates, single letter variable names are dubious. Same for `s` and `t`.

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,613 @@
+  }
+
+  uint32_t p = target_output_indices.size();
+  std::vector<int64_t> s(output_indices.size(), -1), t(output_indices.size(), -1);
```
```suggestion
  std::vector<int64_t> s(output_indices.size(), -1);
  std::vector<int64_t> t(output_indices.size(), -1);
```
Generally avoid hiding additional declarations on the same line.

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,613 @@
+  std::vector<uint32_t> target_output_indices(output_labels.begin(), output_labels.end());
+
+  // map output dim labels to 0 ~ n-1
+  std::vector<uint32_t> arr(output_indices.begin(), output_indices.end());
```
`arr` - what is a clearer name?

fdwr reviewed (commented):
TY Peishen. Some more thoughts. I'll make a few trivial edits directly (little typos), and for the renaming, we can go ahead and submit as-is if you can follow up in a small CR to make some of the local variable names cleaner.

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -0,0 +1,613 @@
+  Step 1. Transpose and Reshape
+
+  (0/1,0/1,0/1) means dim i whether appears in (A,B,C)
+  For new A, it has three segements [...a_1..., a_2, a_3], a_1 has multiple dims, a_2 and a_3 only have one dim respectively
+  For new B, it has three segements [...b_1..., b_2, b_3], b_1 has multiple dims, b_2 and b_3 only have one dim respectively
```
Sorry for the delay, I was awaiting the ONNX change shape inference change to arrive too, but it still has not. We don't need to block on it though.

fdwr commented:
/azp run Windows ARM64 QNN CI Pipeline,Windows x64 QNN CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline

fdwr commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models

fdwr commented:
/azp run Linux Android Emulator QNN CI Pipeline

fdwr commented:
/azp run Windows ARM64 QNN CI Pipeline,Windows x64 QNN CI Pipeline,Windows CPU CI Pipeline,Windows GPU CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline

fdwr commented:
/azp run Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline,Big Models

fdwr commented:
/azp run Linux Android Emulator QNN CI Pipeline

peishenyan commented:
Hi @fdwr, long time no see. Really appreciate your help. I will address the comments today~
BTW, I remember a few months ago when I ran ORT WebNN EP without onnx shape inference, it aborted because the output could not be registered (no output shape). I'm not sure if this branch can be merged before onnx 1.17 is ready, and I will confirm this today.

peishenyan commented:
Hi all, I have addressed the comments in the latest commit. 
However, without onnx v1.17 the ORT with WebNN EP will crash because the output of einsum op is not registered correctly. This branch should not be merged until onnx v1.17 is ready. Thanks.

fdwr commented:
> This branch should not be merged until onnx v1.17 is ready. 

Thanks for confirmation. Ok, will defer merging. So then for any cases that need the shape inference logic (like the Segment Anything encoder https://github.com/microsoft/webnn-developer-preview/blob/main/demos/segment-anything/index.js#L739C57-L739C108), we'll need a custom build of ORT and copy the .wasm files onto the server.

fdwr commented:
Relevant! https://github.com/microsoft/onnxruntime/pull/21897

Honry commented:
https://github.com/microsoft/onnxruntime/pull/22376 this is landed!

@fdwr, Peishen has resolved the conflict and applied latest `OpSupportLimits` feature, pls. take another look, thanks!

tianleiwu commented:
/azp run Windows ARM64 QNN CI Pipeline,Windows x64 QNN CI Pipeline,Windows CPU CI Pipeline,Windows GPU CUDA CI Pipeline,Windows GPU DML CI Pipeline,Windows GPU Doc Gen CI Pipeline,Windows GPU TensorRT CI Pipeline,ONNX Runtime Web CI Pipeline,Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline

tianleiwu commented:
/azp run Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,orttraining-amd-gpu-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,onnxruntime-binary-size-checks-ci-pipeline

tianleiwu commented:
/azp run Big Models,Linux Android Emulator QNN CI Pipeline,Android CI Pipeline,iOS CI Pipeline,ONNX Runtime React Native CI Pipeline

peishenyan commented:
Hi @fdwr, do you have any other comment? Could we merge this branch?

fdwr reviewed (approved):
👍

fdwr commented:
/azp run ONNX Runtime Web CI Pipeline,Windows GPU CI Pipeline,Linux Android Emulator QNN CI Pipeline

fdwr commented:
/azp run Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,Windows ARM64 QNN CI Pipeline,Windows CPU CI Pipeline

fdwr commented:
/azp run Windows GPU CUDA CI Pipeline,Windows GPU DML CI Pipeline,Windows GPU Doc Gen CI Pipeline

fdwr commented:
/azp run Windows GPU TensorRT CI Pipeline,onnxruntime-binary-size-checks-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,Windows x64 QNN CI Pipeline,Big Models

peishenyan commented:
"Windows GPU DML CI pipeline failed, but it does not appear to be related to this PR. I have updated the code to fix the above comment, please re-trigger the CI tests. Thanks!

```
1: [ RUN      ] NhwcMaxPoolContribOpTest.MaxPool2D
1: 2024-11-12 06:36:00.9277083 [W:onnxruntime:Default, base_tester.cc:830 onnxruntime::test::BaseTester::ExecuteModelForEps] registered execution providers DmlExecutionProvider  were unable to run the model.
1: 2024-11-12 06:36:01.1277950 [W:onnxruntime:Default, base_tester.cc:830 onnxruntime::test::BaseTester::ExecuteModelForEps] registered execution providers DmlExecutionProvider  were unable to run the model.
1: 2024-11-12 06:36:01.3399722 [W:onnxruntime:Default, base_tester.cc:830 onnxruntime::test::BaseTester::ExecuteModelForEps] registered execution providers DmlExecutionProvider  were unable to run the model.
1: 2024-11-12 06:36:01.5518211 [W:onnxruntime:Default, base_tester.cc:830 onnxruntime::test::BaseTester::ExecuteModelForEps] registered execution providers DmlExecutionProvider  were unable to run the model.
1: D:\a\_work\1\s\onnxruntime\core\providers\dml\dml_provider_factory.cc(567)\onnxruntime_test_all.exe!00007FF62D76D73A: (caller: 00007FF62D76DAE5) Exception(1) tid(1658) 8007000E Not enough memory resources are available to complete this operation.
1: 
1: Provider:DmlExecutionProvider
1: unknown file: error: C++ exception with description "D:\a\_work\1\s\onnxruntime\core\providers\dml\dml_provider_factory.cc(567)\onnxruntime_test_all.exe!00007FF62D76D73A: (caller: 00007FF62D76DAE5) Exception(1) tid(1658) 8007000E Not enough memory resources are available to complete this operation.
1: " thrown in the test body.
1: 
1: [  FAILED  ] NhwcMaxPoolContribOpTest.MaxPool2D (4516 ms)
```

fdwr commented on onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc:
```diff
@@ -62,7 +62,7 @@ struct Component {
 
 bool ParseEquationComponents(const Node& node,
-                             const std::string& equation,
+                             const std::string_view& equation,
```
String views are passed by value since they are already pointers themselves. Otherwise you're passing a reference to a reference.
```suggestion
                             const std::string_view equation,
```

fdwr commented:
> The following is the result of failed"Windows GPU DML CI pipeline, but it does not appear to be related to this PR.

Hmm, yeah, definitely unrelated.

fdwr commented:
/azp run ONNX Runtime Web CI Pipeline,Windows GPU CI Pipeline,Linux Android Emulator QNN CI Pipeline
/azp run Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,Windows ARM64 QNN CI Pipeline,Windows CPU CI Pipeline
/azp run Windows GPU CUDA CI Pipeline,Windows GPU DML CI Pipeline,Windows GPU Doc Gen CI Pipeline
/azp run Windows GPU TensorRT CI Pipeline,onnxruntime-binary-size-checks-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,Windows x64 QNN CI Pipeline,Big Models

fdwr commented:
/azp run ONNX Runtime Web CI Pipeline,Windows GPU CI Pipeline,Linux Android Emulator QNN CI Pipeline

fdwr commented:
/azp run Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,Windows ARM64 QNN CI Pipeline,Windows CPU CI Pipeline

fdwr commented:
Rerunning failed CI's.

fdwr commented:
I retried the "Linux GPU CI Pipeline", but I still see a failure 🤔.

fdwr commented:
Can you remerge with master to include this? https://github.com/microsoft/onnxruntime/pull/22828

peishenyan commented:
> Can you remerge with master to include this? #22828

Sure

fdwr commented:
/azp run ONNX Runtime Web CI Pipeline,Windows GPU CI Pipeline,Linux Android Emulator QNN CI Pipeline

fdwr commented:
/azp run Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,Windows ARM64 QNN CI Pipeline,Windows CPU CI Pipeline

fdwr commented:
/azp run Windows GPU CUDA CI Pipeline,Windows GPU DML CI Pipeline,Windows GPU Doc Gen CI Pipeline

fdwr commented:
/azp run Windows GPU TensorRT CI Pipeline,onnxruntime-binary-size-checks-ci-pipeline,orttraining-linux-ci-pipeline,orttraining-linux-gpu-ci-pipeline,orttraining-ortmodule-distributed,Windows x64 QNN CI Pipeline,Big Models

fdwr reviewed (approved):
👍
